### PR TITLE
golang: Initial LSP server implementation

### PR DESCRIPTION
**Pull request by keegancsmith**
The implementation is very naive, and only resolves symbols contained in
the current file. However, this is a good starting point for further changes.

### CODE REVIEW

**Comment by keegancsmith** [comment]
Plan going forward:
1. Integrate the VFS solution we end up with.
1. Re-use existing tools that answer these questions (godef, guru). Do this by vendoring + modifying for VFS in such a way that we can maintain the vendored files.
1. Support stretch goals (external references, etc)

cc @rothfels @slimsag

File: lang/golang/hover.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	conf.CreateFromFiles(importPath, f)
+	prog, err := conf.Load()
+	if err != nil {
+		log.Printf("typechecking %s: %s", params.TextDocument.URI, err)
+		if prog == nil {
+			return nil, err
+		}
+	}
+
+	// We already performed the PathEnclosingInterval operation above,
```
Comment (sqs) [inline lang/golang/hover.go]: This comment can be removed. It was for when the implementation called `prog.PathEnclosingInterval(...)`.

File: lang/golang/references.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	conf.CreateFromFiles(importPath, f)
+	prog, err := conf.Load()
+	if err != nil {
+		log.Printf("typechecking %s: %s", params.TextDocument.URI, err)
+		if prog == nil {
+			return nil, err
+		}
+	}
+
+	// We already performed the PathEnclosingInterval operation above,
```
Comment (sqs) [inline lang/golang/references.go]: As mentioned above, this comment can be removed. We no longer call PathEnclosingInterval again.

File: lang/golang/references.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	// performing it again.
+	pkg := prog.InitialPackages()[0]
+	if pkg == nil {
+		return nil, errors.New("no package found")
+	}
+	o := pkg.ObjectOf(node)
+
+	var locs []lsp.Location
+	for astIdent, obj := range pkg.Info.Uses {
+		if sameObj(o, obj) {
```
Comment (sqs) [inline lang/golang/references.go]: Probably worth adding a comment mentioning that this doesn't yet handle includeDeclaration from https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#find-references.

**Comment by sqs** [comment]
LGTM after a couple of minor comments.

**Comment by keegancsmith** [comment]
@sqs: Thanks for the feedback, I'll fix. Just a note: I'll be ripping all this out tomorrow more than likely and replacing it with something far more robust. Just wanted to get a base into master to iterate on.

**Pull request merged**